src/progress.rs

```diff
@@ -141,99 +252,283 @@ impl ProgressSet {
     }
 
     /// Returns an iterator across all the nodes and their progress.
+    ///
+    /// **Note:** Do not use this for majority/quorum calculation. The Raft node may be
+    /// transitioning to a new configuration and have two qourums. Use `has_quorum` instead.
     #[inline]
     pub fn iter(&self) -> impl ExactSizeIterator<Item = (&u64, &Progress)> {
         self.progress.iter()
     }
 
     /// Returns a mutable iterator across all the nodes and their progress.
+    ///
+    /// **Note:** Do not use this for majority/quorum calculation. The Raft node may be
+    /// transitioning to a new configuration and have two qourums. Use `has_quorum` instead.
     #[inline]
     pub fn iter_mut(&mut self) -> impl ExactSizeIterator<Item = (&u64, &mut Progress)> {
         self.progress.iter_mut()
     }
 
     /// Adds a voter node
-    pub fn insert_voter(&mut self, id: u64, mut pr: Progress) -> Result<(), Error> {
-        // If the progress exists already this is in error.
-        if self.progress.contains_key(&id) {
-            // Determine the correct error to return.
-            if self.learner_ids().contains(&id) {
-                return Err(Error::Exists(id, "learners"));
-            }
-            return Err(Error::Exists(id, "voters"));
+    pub fn insert_voter(&mut self, id: u64, pr: Progress) -> Result<(), Error> {
+        debug!("Inserting voter with id {}.", id);
+
+        if self.learner_ids().contains(&id) {
+            Err(Error::Exists(id, "learners"))?;
+        } else if self.voter_ids().contains(&id) {
+            Err(Error::Exists(id, "voters"))?;
         }
-        pr.is_learner = false;
         self.configuration.voters.insert(id);
+        self.next_configuration
```

Comment:
Can we call `insert_voter` if the progress currently in transision?

Reply:
The user can `insert_voter` whenever they want, so it's never safe. They *should* only call it when they find a `EntryConfChange` `Entry`.

Right now the protection for doing so is here:

https://github.com/pingcap/raft-rs/blob/52856d1cdc3d65ec33539a7c9f9cffdd34529e1a/src/raft.rs#L1395-L1401

I am not satisfied with this though, and I definitely support this change.

Reply:
Do we need to restrict the method to `pub(crate) insert_voter`? However, personally I still think just panic here if `next_configuration` is not none is the most simple solution. :)

Reply:
@hicqu How about we make `insert_voter` and such return a Result instead?

Reply:
This would be a quite large change, actually. :( We have a lot of places where we swallow errors.